### PR TITLE
first version GNN

### PR DESCRIPTION
**Pull request by Rodin1000**
1.

### CODE REVIEW

File: cgs_vmc/layers.py [hunk 1]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+  def _build(self, inputs: tf.Tensor) -> tf.Tensor:
+    """Connects the GNN module into the graph, with input `inputs`.
+
+    Args:
+      inputs: Tensor with input values.
+
+    Returns:
+      Result of convolving `inputs` with adjacency matrix and module's kernels.
+    """
+    neighbor_num = len(adj[0])
```
Comment (Rodin1000) [inline cgs_vmc/layers.py]: will this be slow inside tensorflow? If so, will it be better to specify neighbor_sum as an argument in hparams?

File: cgs_vmc/layers.py [hunk 2]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
 
     batch_components = [self._mps_var[:, :, index] for index in index_inputs]
     return tf.stack(batch_components, axis=0)
+
```
Comment (kochkov92) [inline cgs_vmc/layers.py:409]: Use 2 blank lines between top level definitions [link](https://google.github.io/styleguide/pyguide.html?showone=Blank_Lines#35-blank-lines).

File: cgs_vmc/layers.py [hunk 3]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+
+
+class GNN_layer(snt.AbstractModule):
+
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: No blank lines between docstring and `def` statement.

File: cgs_vmc/layers.py [hunk 4]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
 
     batch_components = [self._mps_var[:, :, index] for index in index_inputs]
     return tf.stack(batch_components, axis=0)
+
+
+
+class GNN_layer(snt.AbstractModule):
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: GNN_layer doesn't confront the naming convention [link](https://google.github.io/styleguide/pyguide.html?showone=Blank_Lines#316-naming)

Consider using `GraphConvLayer`

File: cgs_vmc/layers.py [hunk 5]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+  """GNN module with adjacency list"""
+
+  def __init__(
+      self,
+      output_channels: int,
+      adj: np.array,
+      name: str = 'gnn_layer'):
+    """Constructs GNN moduel.
+
+      Args:
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: `Args:` should be aligned with """ of the doc-string at the top.

File: cgs_vmc/layers.py [hunk 6]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+
+
+class GNN_layer(snt.AbstractModule):
+
+  """GNN module with adjacency list"""
+
+  def __init__(
+      self,
+      output_channels: int,
+      adj: np.array,
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: For consistency use `np.ndarray`

File: cgs_vmc/layers.py [hunk 7]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+      name: str = 'gnn_layer'):
+    """Constructs GNN moduel.
+
+      Args:
+      output_channels: Number of output channels in convolution.
+      adj: adjacency list of the graph.
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: Here I would explain the format "Adjacency list that stores indices of neighbors for every site, has shape [n_sites, num_neighors]."

File: cgs_vmc/layers.py [hunk 8]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+      name: Name of the module.
+    """
+    super(GNN_layer, self).__init__(name=name)
+    self._output_channels = output_channels
+    self._adj = adj
+
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: Single blank line between method definitions. [link](https://google.github.io/styleguide/pyguide.html?showone=Blank_Lines#35-blank-lines)

File: cgs_vmc/layers.py [hunk 9]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+    """Connects the GNN module into the graph, with input `inputs`.
+
+    Args:
+      inputs: Tensor with input values.
+
+    Returns:
+      Result of convolving `inputs` with adjacency matrix and module's kernels.
+    """
+    neighbor_num = len(adj[0])
+    adj_layer = tf.gather(inputs,self._adj,axis=1)
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: A single space is required between arguments [link](https://google.github.io/styleguide/pyguide.html?showone=Blank_Lines#36-whitespace) i.e.
`tf.gather(inputs, adjacency_matrix, axis=1)`

p.s. I would also call it `adj_table` or something, but adj_layer also works.

File: cgs_vmc/layers.py (hunk 1 above)
Comment (kochkov92) [reply cgs_vmc/layers.py]: Since this method is building the graph it is executed only once, so there is essentially no overhead in using any methods that are executed during graph construction time. I would look into using `np.shape` for better readability. (I would also name it `num_neighbors` to make sure it's not confused with neighbor_id etc.)

File: cgs_vmc/layers.py [hunk 10]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+    conv = tf.layers.conv2d(adj_layer, self._filters, kernel_size=(1,neighbor_num))
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: Similar to above `kernel_size=(1, neighbor_num)`;

Also this line is too long, break it in 2 lines.

File: cgs_vmc/run_supervised_training.py [hunk 11]
```diff
@@ -55,6 +55,9 @@
 flags.DEFINE_string(
     'basis_file_path', '',
     'Path to the basis file for full wavefunction evaluation.')
+flags.DEFINE_string(
```
Comment (kochkov92) [inline cgs_vmc/run_supervised_training.py]: Adding a GraphConvolutionalNetwork specific flag to a general pipeline makes code harder to read. Please delegate the parameters to `hparams` (i.e. remove this and consider passing the adjacency_matrix_file_name in hparams).

File: cgs_vmc/run_training.py [hunk 12]
```diff
@@ -53,6 +53,9 @@
 flags.DEFINE_string(
     'basis_file_path', '',
     'Path to the basis file for full wavefunction evaluation.')
+flags.DEFINE_string(
```
Comment (kochkov92) [inline cgs_vmc/run_training.py]: Same as in run_supervised_training, we don't want to put one model specific parameters in a general pipeline.

File: cgs_vmc/utils.py [hunk 13]
```diff
@@ -107,6 +107,9 @@ def create_hparams(**kwargs: Any) -> tf.contrib.training.HParams:
       bot_lin_table_file='',
       ed_vector_file='',
 
+      # GNN parameters
+      adj_list = 'adj.txt',
```
Comment (kochkov92) [inline cgs_vmc/utils.py]: I would leave this blank: reasoning is that if you mess up the input, you prefer your code to scream that file is not found, rather than silently giving your wrong results.

I would also specify that this is `adjacency_list_file_name`

File: cgs_vmc/layers.py [hunk 14]
```diff
@@ -405,3 +406,40 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+    Args:
+      inputs: Tensor with input values.
+
+    Returns:
+      Result of convolving `inputs` with adjacency matrix and module's kernels.
+    """
+    neighbor_num = len(adj[0])
+    adj_layer = tf.gather(inputs,self._adj,axis=1)
+    conv = tf.layers.conv2d(adj_layer, self._filters, kernel_size=(1,neighbor_num)) 
+    return tf.squeeze(conv,axis=2)
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: `tf.squeeze(conv, axis=2)`

File: cgs_vmc/wavefunctions.py [hunk 15]
```diff
@@ -1080,6 +1080,85 @@ def from_hparams(
...
 
 
+class GNN(Wavefunction):
+  """Implementation of wavefunction as graph neural network."""
+
+  def __init__(
+      self,
+      num_layers: int,
+      num_filters: int,
+      adj: np.array,
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: `np.ndarray`

File: cgs_vmc/wavefunctions.py [hunk 16]
```diff
@@ -1080,6 +1080,85 @@ def from_hparams(
     return cls(**full_vector_params)
 
 
+class GNN(Wavefunction):
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: I think this could be more descriptive, considering that this is not a well-known architecture. e.g. `GraphConvNetwork`  (this also avoids abbreviations in class names, which is a plus)

File: cgs_vmc/wavefunctions.py [hunk 17]
```diff
@@ -1080,6 +1080,85 @@ def from_hparams(
...
+  ) -> tf.Tensor:
+    """Builds computational graph evaluating the wavefunction on inputs.
+
+    Args:
+      inputs: Input tensor, must have shape (batch, num_sites, ...).
+
+    Returns:
+      Tensor holding values of the wavefunction on `inputs`.
+
+    Raises:
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: Doc-string should not documents exceptions that you do not raise.

File: cgs_vmc/wavefunctions.py [hunk 18]
```diff
@@ -1080,6 +1080,85 @@ def from_hparams(
...
+    gnn_params = {
+        'num_layers': hparams.num_conv_layers,
+        'num_filters': hparams.num_conv_filters,
+        'adj': np.genfromtxt(hparams.adj_list, dtype=int),
+        'output_activation': layers.NONLINEARITIES[hparams.output_activation],
+        'nonlinearity': layers.NONLINEARITIES[hparams.nonlinearity],
+    }
+    if name:
+      gnn_params['name'] = name
+    return cls(**gnn)
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: this code doesn't work, I think you meant `**gnn_params`

File: cgs_vmc/wavefunctions.py [hunk 19]
```diff
@@ -1080,6 +1080,85 @@ def from_hparams(
...
+
+
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: This should be 2 blank lines between top level definitions.

**Review by kochkov92 — CHANGES_REQUESTED** [review]
First portion of comments.

**Comment by Rodin1000** [comment]
Thanks for the comment. Here's the update.

File: cgs_vmc/layers.py [hunk 20]
```diff
@@ -405,3 +406,38 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+  def __init__(
+      self,
+      output_channels: int,
+      adj: np.ndarray,
+      name: str = 'gnn_layer'):
+    """Constructs GraphConvLayer moduel.
+
+    Args:
+      output_channels: Number of output channels in convolution.
+      adj: Adjacency list of the graph that stores indices of neighbors and itself for every site, with shape [n_site, num_neighbor].
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: Maximum line length by many standards is 80 characters. (See Conv2dPeriodic for an example of breaking doc-string into multiple lines)

File: cgs_vmc/layers.py [hunk 21]
```diff
@@ -405,3 +406,38 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+    Args:
+      inputs: Tensor with input values.
+
+    Returns:
+      Result of convolving `inputs` with adjacency matrix and module's kernels.
+    """
+    num_neighbor = np.shape(adj)[1]
+    adj_table = tf.gather(inputs, self._adj, axis=1)
+    conv = tf.layers.conv2d(adj_table, self._filters, \
+            kernel_size=(1, num_neighbor))
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: As per style guide, do not use backslash line continuation except for `with` statement [link](http://google.github.io/styleguide/pyguide.html#32-line-length)

For splitting long lines see [examples](http://google.github.io/styleguide/pyguide.html#34-indentation)

Also many style guides suggest removing trailing whitespace (this can be easily automated to be formatted on save/write)

File: cgs_vmc/layers.py [hunk 22]
```diff
@@ -405,3 +406,38 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+  def _build(self, inputs: tf.Tensor) -> tf.Tensor:
+    """Connects the GraphConvLayer module into the graph, with input `inputs`.
+
+    Args:
+      inputs: Tensor with input values.
+
+    Returns:
+      Result of convolving `inputs` with adjacency matrix and module's kernels.
+    """
+    num_neighbor = np.shape(adj)[1]
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: This actually doesn't work, as `adj` is not defined in this scope. Use the class attribute `self._adj`.

File: cgs_vmc/wavefunctions.py [hunk 23]
```diff
@@ -1080,6 +1080,80 @@ def from_hparams(
...
+  """Implementation of wavefunction as graph neural network."""
+
+  def __init__(
+      self,
+      num_layers: int,
+      num_filters: int,
+      adj: np.ndarray,
+      nonlinearity: tf.Tensor = tf.nn.relu,
+      output_activation: tf.Tensor = tf.exp,
+      name: str = 'graph_neural_network',
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: nit: As suggested by Sonnet library, default names of snt.AbstractModule should be  equal to class name converted from CamelCase to underscore case.  i.e. in this case it would be `graph_conv_network `.

File: cgs_vmc/layers.py [hunk 24]
```diff
@@ -405,3 +406,38 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+
+
+class GraphConvLayer(snt.AbstractModule):
+  """GraphConvLayer module with adjacency list"""
+
+  def __init__(
+      self,
+      output_channels: int,
+      adj: np.ndarray,
+      name: str = 'gnn_layer'):
```
Comment (kochkov92) [inline cgs_vmc/layers.py]: nit: As suggested by Sonnet library, default names of snt.AbstractModule should be  equal to class name converted from CamelCase to underscore case.  i.e. in this case it would be `graph_conv_layer `.

File: cgs_vmc/wavefunctions.py [hunk 25]
```diff
@@ -1080,6 +1080,80 @@ def from_hparams(
...
+  ):
+    """Creates an instance of a class.
+
+    Args:
+      num_layers: Number of convolutional layers.
+      num_filters: Number of convolutional filters in each layer.
+      adj: adjacency list of the graph.
```
Comment (kochkov92) [inline cgs_vmc/wavefunctions.py]: For consistency start docstring with a capital letter: `Adjacency ...`.

File: cgs_vmc/utils.py [hunk 26]
```diff
@@ -107,6 +107,9 @@ def create_hparams(**kwargs: Any) -> tf.contrib.training.HParams:
       bot_lin_table_file='',
       ed_vector_file='',
 
+      # GraphConvNetwork parameters
+      adjacency_list = '',
```
Comment (kochkov92) [inline cgs_vmc/utils.py]: As per style guide [link](http://google.github.io/styleguide/pyguide.html#36-whitespace) whitespace is not used for default values unless type annotation is used.

`adjacency_list='',`

**Review by kochkov92 — CHANGES_REQUESTED** [review]
A few more comments.

**Comment by Rodin1000** [comment]
Here's the update, please just let me know if there's any suggestion.

File: cgs_vmc/utils.py [hunk 27]
```diff
@@ -107,6 +107,9 @@ def create_hparams(**kwargs: Any) -> tf.contrib.training.HParams:
       bot_lin_table_file='',
       ed_vector_file='',
 
+      # GraphConvNetwork parameters
+      adjacency_list='',
```
Comment (kochkov92) [inline cgs_vmc/utils.py]: I would personally add a specifier to the name that this is a file, not an adjacency_list.

It is also nice to put a vert short description of the parameter above (in the docstring)

File: cgs_vmc/layers.py [hunk 28]
```diff
@@ -405,3 +406,42 @@ def _build(self, inputs: tf.Tensor) -> tf.Tensor:
...
+          and itself for every site, with shape [n_site, num_neighbor].
+      name: Name of the module.
+    """
+    super(GraphConvLayer, self).__init__(name=name)
+    self._output_channels = output_channels
+    self._adj = adj
+    num_neighbors = np.shape(self._adj)[1]
+    kernel_shape = (1, num_neighbors)
+    with self._enter_variable_scope():
+      self._conv_module = snt.Conv2D(output_channels=output_channels,
```
Comment (kochkov92) [inline cgs_vmc/layers.py:433]: Replaced tf.layers.conv with snt.Conv2D module.

**Review by kochkov92 — APPROVED** [review]
Please check that my small edit didn't break anything. I marked some modifications in the comments.

P.S. We need to modify your script for generation of adjacency file, as it does not preserve the order in which neighbors are listed, which breaks translational invariance. As of my commit the 1D analog works as expected.
